packages/lib/src/ecdsa.ts

```diff
@@ -0,0 +1,70 @@
+const ECSignature = require("elliptic/lib/elliptic/ec/signature");
+const BN = require("bn.js");
+import { WeierstrassPoint, babyjubjub } from "./babyJubjub";
+import { Signature } from "./types";
+import { publicKeyFromString } from "./utils";
+
+export const privateKeyToPublicKey = (privKey: bigint): WeierstrassPoint => {
+  const pubKeyPoint = babyjubjub.ec.g.mul(privKey.toString(16));
+  return WeierstrassPoint.fromEllipticPoint(pubKeyPoint);
+};
+
+export const verifyEcdsaSignature = (
+  sig: Signature,
+  msgHash: bigint,
+  pubKey: WeierstrassPoint
+): boolean => {
+  const ecSignature = new ECSignature({
+    r: sig.r.toString(16),
+    s: sig.s.toString(16),
+  });
+  const pubKeyPoint = babyjubjub.ec.curve.point(
+    pubKey.x.toString(16),
+    pubKey.y.toString(16)
+  );
+  const ecPubKey = babyjubjub.ec.keyFromPublic(pubKeyPoint);
+
+  return babyjubjub.ec.verify(msgHash, ecSignature, ecPubKey);
+};
+
+export const recoverPubKeyIndexFromSignature = (
+  sig: Signature,
+  msgHash: bigint,
+  pubKeys: string[]
+): number => {
+  const Fb = babyjubjub.Fb;
+  const Fs = babyjubjub.Fs;
+
+  const pubKeyEdwardsList = pubKeys.map((pubKey) => {
+    const pubKeyWeierstrass = publicKeyFromString(pubKey);
+    return pubKeyWeierstrass.toEdwards();
+  });
+
+  for (let i = 0; i < babyjubjub.cofactor; i++) {
```

Comment:
I don't think I understand why we need to have this outer for loop. Won't the signatures be living in the subgroup generated by the base point? This for loop seems to be trying versions of R that are in different subgroups, but won't we know which one we need to use based on our base point selection?

Reply:
We need to loop because r is modded by the prime subgroup order, and this loop allows us to recover all possible original values